modules/oe_webtools_analytics/modules/oe_webtools_analytics_rules/src/RuleMatcher.php

```diff
@@ -0,0 +1,273 @@
+use Drupal\Core\Path\AliasManagerInterface;
+use Drupal\Core\Path\CurrentPathStack;
+use Drupal\oe_webtools_analytics_rules\Entity\WebtoolsAnalyticsRule;
+use Drupal\oe_webtools_analytics_rules\Entity\WebtoolsAnalyticsRuleInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Service that matches routes to rules.
+ */
+class RuleMatcher implements RuleMatcherInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The request stack.
+   *
+   * @var \Symfony\Component\HttpFoundation\RequestStack
+   */
+  protected $requestStack;
+
+  /**
+   * The current path service.
+   *
+   * @var \Drupal\Core\Path\CurrentPathStack
+   */
+  protected $currentPath;
+
+  /**
+   * The alias manager service.
+   *
+   * @var \Drupal\Core\Path\AliasManagerInterface
+   */
+  protected $aliasManager;
+
+  /**
+   * A cache backend interface.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cache;
+
+  /**
+   * The configuration object factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $config;
+
+  /**
+   * Constructs a RuleMatcher service.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
+   *   The request stack.
+   * @param \Drupal\Core\Path\CurrentPathStack $currentPath
+   *   The current path service.
+   * @param \Drupal\Core\Path\AliasManagerInterface $aliasManager
+   *   The alias manager service.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache
+   *   A cache backend used to store webtools rules for uris.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config
+   *   The Config Factory service.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, RequestStack $requestStack, CurrentPathStack $currentPath, AliasManagerInterface $aliasManager, CacheBackendInterface $cache, ConfigFactoryInterface $config) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->requestStack = $requestStack;
+    $this->currentPath = $currentPath;
+    $this->aliasManager = $aliasManager;
+    $this->cache = $cache;
+    $this->config = $config;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getMatchingSection(string $path = NULL): ?string {
+    $current_path = $this->getCurrentPath();
+    $cache = $this->cache->get($current_path) ?: new \stdClass();
+
+    // Check if the cache data for the section has been set, taking into account
+    // that its value might be NULL.
+    if (empty($cache->data) || !array_key_exists('section', $cache->data)) {
+      $rule = $this->getMatchingRule($current_path);
+
+      // Reload the cache object, it is updated when the rule was matched.
+      $cache = $this->cache->get($current_path) ?: new \stdClass();
```

Comment:
I had to add it because the tests coverage pointed out that the cache items were not correctly invalidated if a rule is matching on the site default language and the default language is being changed. It needs to have the cache tags added of the default site configuration, but only if the rule matches on the site default language.

The cache entry is not per section or per rule, but per path, so this is fine. A path entry has two data points: the section to which the path belongs and the rule that is responsible for setting it.

I'll have a look to refactor this so that the weird cache juggling is not being done. It will be cleaner and simpler if we just populate both on a cache miss.

Reply:
Washed the yaks.